### DISCUSSION TITLE
Hiding categories (and maybe reorganizing them)

### DISCUSSION THREAD

**Discussion by sfederici** [ScratchAddons/ScratchAddons#6869] [Ideas] [answered]
Adding new "official" blocks to Scratch categories -the same holds for TW- is a process that (even if, in the end, I succeded almost an year ago) is not simple and straightforward at all, especially because doing it both for Windows and macOS requires several adjustments. 

I noticed in TW extensions the strategy of creating "plus" categories (for example "variable and list") where all the standard blocks are recreated and new blocks are added. If standard categories could be hidden, this mechnism could be used to create new blocks in standard categories too, by recreating all the standard blocks (pretty easy and straightforward) and adding the new blocks.

It would be also very useful, if doable, if categories and extensions could be reorganized in the desired order, so that the "plus" category handled by the extension could fully replace the standard category.

**Comment by WorldLanguages** [comment]
I don't think I understood your post, could you provide more context? Does this actually relate to Scratch Addons, or just TurboWarp extensions? (even if it's TurboWarp, I'm surprised development needs to consider runtime support for multiple operating systems)

**Reply by sfederici** [reply]
My needs could be both solved by either ScratchAddons or Turbowarp.

I'm currently using a customized version of Turbowarp Desktop, as I need blocks that are not available in Scratch (but that were in the past, so I'm adding them to a standard category) but I also need a lot of addons. If I could hide (a  subset of) the standard categories, I could just use addons and extensions -without need for a customized Turbowarp. If I could also rearrange the items in the category/extension selector, that would be the perfect solution for my problem: I could "simulate" an updated standard category by hiding the standard category and repositioning an extension identical to the standard category.

If such an addon could not be realized, but the possibility to hide standard categories could become a Turbowarp feature, that would be another way to solve my problem.

Let me know if this is not fully clear yet.

**Reply by WorldLanguages** [reply]
It would be nice to see a screenshot of your custom TurboWarp version.

**Reply by sfederici** [reply]
Nothing fancy. I just added back the Scratch 1.4. FOREVER IF block to the Control category. I could have added it via a custom extension, but I wanted it to be a "true" Control block, as it was in Scratch 1.4. I hadn't thought that I could recreate the whole set of Control blocks in an extension and give them the same color of the Control category. But then, the standard Control category would still be there. And that wouldn't be a clean solution to my need.

**Reply by CST1229** [reply]
I think it's possible to override categories using unsandboxed custom extensions (not sure how but it involves creating an extension with the same ID as an existing category (EDIT: actually it is just that)). Also, you can use Scratch.BlockType.XML (added by TW) to add existing blocks to an extension category (for an example, see the Hacked Block Collection extension in the extensions.turbowarp.org library).

**Comment by sfederici** [comment]
> I think it's possible to override categories using custom extensions (not sure how but it involves creating an extension with the same ID as an existing category (EDIT: actually it is just that)

I will certainly give it a try. I didn't think Scratch developers would have allowed to overwrite their standard blocks so easily. Good for me.

> Also, you can use Scratch.BlockType.XML to add existing blocks to an extension category (for an example, see the Hacked Block Collection extension in the extensions.turbowarp.org library)

That would make my job even easier. Thanks a lot for sharing this information.

**Reply by sfederici** [reply]
I hope I didn't study the whole extentions' documentation to do something that I can't do. What I need to create indeed is a new C-SHAPED block (the old "forever if" available in Scratch 1.4). But, in the block types, in the whole documentation and in the extensions in the gallery I couldn't find any example of such blocks. Are C-SHAPED blocks something that cannot be created in Scratch/TurboWarp extensions? Or did I simply miss the relevant steps? Thanks in advance for your help

**Reply by CST1229** [reply, marked as answer]
They aren't very documented I think as TurboWarp's compiler didn't support them until somewhat recently, but there are loop BlockTypes: `Scratch.BlockType.CONDITIONAL` for C blocks that can run their contents once, and `Scratch.BlockType.LOOP` for loop blocks that can run their contents any number of times (and have a loop arrow). You can also specify a `branchCount` property to have more than 1 branch in the block (this defaults to 1 for C blocks).
In the block function you can then either return the number of the branch to run (starting from 1, or a falsy value to not run the branch and end the loop for loop blocks), or use `util.startBranch` like in core blocks.

**Reply by sfederici** [reply]
Success!!! ([look here](https://github.com/TurboWarp/extensions/issues/1162#issuecomment-1825973449))

**Reply by WorldLanguages** [reply]
Thanks CST!
